### PR TITLE
Close a listing axel

### PR DESCRIPTION
**Pull request by axelnguyen0701**
- Close an instant listing done (tested under "Listing C.R.U.D")
- Close a bidding listing done (tested under "Bidding")

### CODE REVIEW

File: contracts/BookMarketplace.sol [hunk 1]
```diff
@@ -290,10 +309,76 @@ contract BookMarketplace is ERC721URIStorage {
...
+    function createBiddingSale(
+        uint256 tokenId
+    ) public onlyListingSeller(tokenId) {
+        require(
+            idToMarketItem[tokenId].allowBid,
+            "This listing does not allow bidding"
+        );
+
+        int256 highestBidIndex = getHighestBidIndex(tokenId);
+        require(highestBidIndex > 0, "This listing does not have any bid yet");
```
Comment (sniezekk) [inline contracts/BookMarketplace.sol:356]: As there is an empty bid initialized in the list, will this ever happen?

File: contracts/BookMarketplace.sol [hunk 2]
```diff
@@ -290,10 +309,76 @@ contract BookMarketplace is ERC721URIStorage {
...
+        require(bid.bidder == msg.sender, "You are not the bidder");
+        require(
+            bid.bidAmount == msg.value,
+            "Please submit the price equals the bid amount"
+        );
+
+        //Transfer bid price
+        address payable seller = idToMarketItem[tokenId].seller;
+        (bool success, ) = seller.call{value: msg.value}("");
+        require(success, "Transfer failed.");
```
Comment (sniezekk) [inline contracts/BookMarketplace.sol:377]: As an aside: we should probably figure out some kind of callback function that is appropriate for using .call, at some point. (no follow-up needed)

File: contracts/BookMarketplace.sol [hunk 3]
```diff
@@ -290,10 +309,76 @@ contract BookMarketplace is ERC721URIStorage {
...
+        //Transfer token
+        idToMarketItem[tokenId].owner = payable(bid.bidder);
```
Comment (sniezekk) [inline contracts/BookMarketplace.sol]: These steps seem pretty similar to the steps in createMarketSale. Is there any value in creating a helper function here?

**Review by sniezekk — COMMENTED** [review]
Pretty good overall, left a few questions.

I think that there is likely some value in trying to write tests to cover the negative cases of your new functions, just so that we know it will fail in the right way.

File: contracts/BookMarketplace.sol (hunk 1 above)
Comment (axelnguyen0701) [reply contracts/BookMarketplace.sol:356]: that empty bid will always have an index of 0, and as highestBidIndex must be > 0, the requirement will only pass if there is an "actual" bid.

File: contracts/BookMarketplace.sol (hunk 3 above)
Comment (axelnguyen0701) [reply contracts/BookMarketplace.sol]: You mean from line 369 onwards? I think we could create a helper function for that

Comment (axelnguyen0701) [reply contracts/BookMarketplace.sol]: Helper function is up.

File: contracts/BookMarketplace.sol (hunk 1 above)
Comment (sniezekk) [reply contracts/BookMarketplace.sol:356]: Oh, duh. My brain also seems to be toast today haha.

File: contracts/BookMarketplace.sol [hunk 4]
```diff
@@ -290,10 +309,82 @@ contract BookMarketplace is ERC721URIStorage {
...
+        //Transfer funds
+        address payable seller = idToMarketItem[tokenId].seller;
+        (bool success, ) = seller.call{value: msg.value}("");
+        require(success, "Transfer failed.");
+
+        //Transfer token
+        transferToken(tokenId);
+
+        //Delete bid history
+        for (uint i = 0; i < idToMarketItem[tokenId].bidList.length; i++) {
```
Comment (sniezekk) [inline contracts/BookMarketplace.sol:383]: Why do we want to delete the bid history?

**Review by sniezekk — APPROVED** [review]

**Pull request merged**